Mathlib/Tactic/Linter/UnusedInstancesInType.lean

```diff
@@ -0,0 +1,222 @@
+optimization. However, the `Parameter`s are created in a telescope, and their fields will *not*
+have loose bound variables.
+-/
+private def _root_.Lean.ConstantVal.onUnusedInstancesWhere (decl : ConstantVal)
+    (p : Expr → Bool) (logOnUnused : Array Parameter → TermElabM Unit) : CommandElabM Unit := do
+  let unusedInstances := decl.type.getUnusedForallInstanceBinderIdxsWhere p
+  if let some maxIdx := unusedInstances.back? then liftTermElabM do
+    unless decl.type.hasSorry do -- only check for `sorry` in the "expensive" case
+      forallBoundedTelescope decl.type (some <| maxIdx + 1)
+        (cleanupAnnotations := true) fun fvars _ => do
+          let unusedInstances : Array Parameter ← unusedInstances.mapM fun idx =>
+            return {
+                fvar? := fvars[idx]?
+                type? := ← fvars[idx]?.mapM (inferType ·)
+                idx
+              }
+          logOnUnused unusedInstances
+
+/--
+Finds theorems whose bodies were elaborated in the current infotrees and whose (full)
+declaration names satisfy `nameFilter`. Checks their type to see if it contains instance hypotheses
+that (1) are unused in the remainder of the type (2) have types which satisfy `instanceTypeFilter`.
+(Note: `instanceTypeFilter` is non-monadic, and may encounter bound variables in its argument. This
+is a performance optimization. `isAppOrForallOfConstP` may be useful in detecting constant
+applications and types of the form `∀ (...), bar ..` here.)
+
+If any such parameters are found in the type of a theorem `foo`, we create a telescope in which the
+types and free variables of the unused parameters are available as
+`unusedParams : Array Parameter := #[p₁, p₂, ..., pₙ]`, as well as the theorem `thm : ConstantVal`
+and current infotree `t`, and run `log t thm unusedParams`.
+
+A simple pattern is therefore
+```
+fun _ thm unusedParams => do
+  logLint linter.fooLinter (← getRef) m!"\
+    {thm.name.unusedInstancesMsg unusedParams}\n\
+    <extra caption>"
+```
+which logs
+
+> \`{foo}\` has the hypothes[is/es] {p₁}, {p₂}, ..., and {pₙ} which [is/are] not used in the
+  remainder of the type.
+>
+> \<extra caption\>
+>
+> This linter can be disabled with \`set_option {linter.fooLinter.name} false\`
+
+pluralizing as appropriate.
+-/
+@[nolint unusedArguments] -- TODO: we plan to use `_cmd` in future
+def _root_.Lean.Syntax.logUnusedInstancesInTheoremsWhere (_cmd : Syntax)
+    (instanceTypeFilter : Expr → Bool)
+    (log : InfoTree → ConstantVal → Array Parameter → TermElabM Unit)
+    (declFilter : ConstantVal → Bool := fun _ => true) : CommandElabM Unit := do
+  for t in ← getInfoTrees do
+    let thms := t.getTheorems (← getEnv) |>.filter declFilter
+    for thm in thms do
+      thm.onUnusedInstancesWhere instanceTypeFilter fun unusedParams =>
+        -- TODO: restore in order to log on type signature
+        -- t.withDeclSigRef cmd thm.name do
```

Comment:
What is the intention here? Is this meant to be a persistent TODO? If so, can you make it clearer what needs doing?